runner/ollamarunner/cache.go

```diff
@@ -235,15 +235,22 @@ func countCommonPrefix(a []*input.Input, b []*input.Input) int32 {
 	return count
 }
 
-// TODO(jessegross): If we need to reprocess the inputs we should ensure that
-// we don't split up a SameBatch
-func (c *InputCache) ShiftDiscard(inputLen int32, numKeep int32) int32 {
-	targetFree := (c.numCtx - numKeep) / 2
-	targetFree = max(targetFree, 1)
-
-	currentFree := c.numCtx - inputLen
+// ShiftDiscard computes how many inputs can be discarded from the cache. Inputs in the same batch
+// are discarded together.
+func (c *InputCache) ShiftDiscard(inputs []*input.Input, numKeep int32) int32 {
+	targetFree := max((c.numCtx-numKeep)/2, 1)
+	currentFree := c.numCtx - int32(len(inputs))
+
+	var discard int32
+	for discard < max(targetFree-currentFree, 0) {
+		if sameBatch := inputs[numKeep+discard].SameBatch; sameBatch > 0 {
+			discard += int32(sameBatch)
```

Comment:
`SameBatch` is the number of tokens following the current one that need to be in the same batch so I believe that this should be `discard += 1 + int32(sameBatch)`. You actually should not need to special case it - if SameBatch is 0 then it the same as the current non-SameBatch case.

The behavior of this loop is a little bit different from how we do truncation in NewSequence, which is SameBatch aware. That one will keep extending discard is there are overlapping SameBatch. That scenario is sort of undefined behavior but it's better to be consistent about it.

Reply:
It looks like some models probably use SameBatch that way but others use the original definition. Regardless, the runner has always executed batches as described above, so that's how these models are being run. It doesn't help things for shifting to have a different interpretation from the rest of the runner.